posts/building-a-realtime-app-with-phoenix.md

```diff
@@ -0,0 +1,604 @@
+
+1. The client will send a join request. 
+2. The join request will get routed by `UserSocket` to the `ChallengesChannel`. 
+3. The `ChallengesChannel` will invoke it's own `join` function which will find the challenge with the given ID in the database and return it to the client. 
+
+![](../images/real-time-phoenix-image-two.svg) 
+
+Let's take a look at the first step, sending the join request from the client. 
+
+#### Sending the Join Request from the Client
+
+```javascript
+import { Socket }   from 'phoenix';
+
+...
+const channel = socket.channel(`challenges:${challengeId}`);
+channel.join().receive('ok', (responsePayload) => {
+  // do something like display the content of responsePayload.challenge
+});
+```
+
+Here we've joined a challenge channel with a topic of the given challenge ID. This sends the request that gets routed via the `UserSocket` to the `ChallengesChannel`, which in turn invokes its `join` function. Let's take a look at this next step now.
+
+#### Handling the Join Request on the Server
+
+```elixir
+defmodule PhoenixPair.ChallengeChannel do
+  use PhoenixPair.Web, :channel
+
+  def join("challenges:" <> challenge_id, _params, socket) do
+    challenge = get_challenge(challenge_id)
+    socket    = assign(socket, :challenge, challenge)
+    response  = %{challenge: challenge}
+    {:ok, response, socket}
+  end
+```
+
+The `join` function does a few things for us:
+* Fetches the challenge with the given ID via the `get_challenge` function (not shown here). 
+* Stores the current challenge in the `socket`'s state via a call to `assign`. 
+* Constructs the response body to send back to the client
+* Responds to the client with `{:ok, response, socket}`
+
+Our client can then use the response body to display the current challenge to the user who just joined this challenge channel.
+
+Now that the user has joined the channel, we're ready to build out the next facet of our real-time feature: a user typing into the shared text editor.
+
+### Step 4: Broadcasting Messages
+
+When a user types into the shared text editor for a given code challenge, we need to do three things:
+
+1. Send a message from the client to the server via the channel.
+2. Teach the server to respond to this message by: updating the challenge's response in the database and broadcasting the updated challenge message to all of the subscribers. 
+3. Teach the subscribing clients to handle the broadcast of the message.
+
+Our overall code flow will go something like this:
+
+![](../images/real-time-phoenix-image-three.svg)
+
+
+#### Sending a Message From the Client
+
+We need to push a message to our channel whenever someone types into the text editor to update the challenge's response. We can do that on the client side by listening to keystrokes in the text area and invoking the following:
+
+```javascript
+// where codeResponse is the updated content of the text area
+channel.push("response:update", {response: codeResponse})
```

Comment:
You don't have to do this, but I think it would be clearer to wrap this code in something like:

```javascript
let onchange_handler = function (codeResponse) {
  channel.push("response:update", {response: codeResponse})
}
```